In PWA, HTML Video Element may be unable to play stream from 'getUserMedia()'

youennf opened the pull request:
#### bbf602136b8138ff22216c8ebc3b0a6c4636fc73
<pre>
In PWA, HTML Video Element may be unable to play stream from &apos;getUserMedia()&apos;
<a href="https://bugs.webkit.org/show_bug.cgi?id=252465">https://bugs.webkit.org/show_bug.cgi?id=252465</a>
rdar://111500785

Reviewed by Eric Carlson.

In case the same GPU process is used by several SafariViewServices, mediaserverd may think that
a capturing session is in the background while it is in foreground.
Ideally mediaserverd would compute this from an audit token or something similar we would pass them.

Currently, it is based on pid forwarding.
As a temporary workaround, we are now storing which pid is to be used for a given WebProcess in RemoteMediaSessionHelperProxy.
When starting to produce data for camera, we are then ensuring that this pid is correctly set whenever starting to produce data.

We allow overriding of pid in MediaSessionHelperiOS, by adding a new ShouldOverride boolean.
Only AVVideoCaptureSource is using ShouldOverride::Yes, which should limit potential fallouts to camera capture cases.

* Source/WebCore/platform/audio/ios/MediaSessionHelperIOS.h:
* Source/WebCore/platform/audio/ios/MediaSessionHelperIOS.mm:
(MediaSessionHelperiOS::providePresentingApplicationPID):
* Source/WebKit/GPUProcess/GPUConnectionToWebProcess.cpp:
(WebKit::GPUConnectionToWebProcess::overridePresentingApplicationPIDIfNeeded):
* Source/WebKit/GPUProcess/GPUConnectionToWebProcess.h:
* Source/WebKit/GPUProcess/media/ios/RemoteMediaSessionHelperProxy.cpp:
(WebKit::RemoteMediaSessionHelperProxy::providePresentingApplicationPID):
(WebKit::RemoteMediaSessionHelperProxy::overridePresentingApplicationPIDIfNeeded):
* Source/WebKit/GPUProcess/media/ios/RemoteMediaSessionHelperProxy.h:
* Source/WebKit/UIProcess/Cocoa/UserMediaCaptureManagerProxy.cpp:
(WebKit::UserMediaCaptureManagerProxy::startProducingData):
* Source/WebKit/UIProcess/Cocoa/UserMediaCaptureManagerProxy.h:
(WebKit::UserMediaCaptureManagerProxy::ConnectionProxy::startProducingData):
* Source/WebKit/WebProcess/GPU/media/ios/RemoteMediaSessionHelper.cpp:
(WebKit::RemoteMediaSessionHelper::providePresentingApplicationPID):
* Source/WebKit/WebProcess/GPU/media/ios/RemoteMediaSessionHelper.h:

Canonical link: <a href="https://commits.webkit.org/265986@main">https://commits.webkit.org/265986@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/30c22a975f38bca09378f90962e2a4b1a097f459

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/11413 "1 style error") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/11617 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/11968 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/13058 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/10863 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/11434 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/14002 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/11599 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/13775 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/11576 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/12447 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/9662 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/13478 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/9739 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/10351 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/17519 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/10810 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/10505 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/13708 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/10913 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/8986 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/10088 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/3068 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/14361 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/10769 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

jernoble commented on Source/WebKit/WebProcess/GPU/media/ios/RemoteMediaSessionHelper.cpp:
```diff
@@ -72,8 +72,10 @@ void RemoteMediaSessionHelper::stopMonitoringWirelessRoutesInternal()
+void RemoteMediaSessionHelper::providePresentingApplicationPID(int pid, ShouldOverride shouldOverride)
 {
+    UNUSED_PARAM(shouldOverride);
+    ASSERT(shouldOverride == ShouldOverride::No);
```
This can be `ASSERT_UNUSED(shouldOverride, shouldOverride == ShouldOverride::No);`

youennf replied:
Will fix.

youennf replied:
Done

jernoble commented on Source/WebKit/GPUProcess/media/ios/RemoteMediaSessionHelperProxy.cpp:
```diff
@@ -32,6 +32,7 @@
 #include "GPUConnectionToWebProcess.h"
 #include "RemoteMediaSessionHelperMessages.h"
 #include "WebCoreArgumentCoders.h"
+#include <WebCore/RuntimeApplicationChecks.h>
```
Is this necessary?

youennf replied:
Actually not currently, will remove.

youennf replied:
Done

eric-carlson reviewed (approved):
r=me

jernoble commented on Source/WebCore/platform/mediastream/mac/AVVideoCaptureSource.h:
```diff
@@ -158,6 +158,10 @@ class AVVideoCaptureSource : public RealtimeVideoCaptureSource, private Orientat
     uint64_t m_lastFramesCount { 0 };
+
+#if PLATFORM(IOS) || PLATFORM(VISION)
+    std::optional<int> m_presentingApplicationPID;
```
This seems like the wrong place to store this value. WE should be storing the presenting PID in the GPU process, on a per GPUToWebContentConnection basis.

youennf replied:
Done

youennf committed:
In PWA, HTML Video Element may be unable to play stream from 'getUserMedia()'
https://bugs.webkit.org/show_bug.cgi?id=252465
rdar://111500785

Reviewed by Eric Carlson.

In case the same GPU process is used by several SafariViewServices, mediaserverd may think that
a capturing session is in the background while it is in foreground.
Ideally mediaserverd would compute this from an audit token or something similar we would pass them.

Currently, it is based on pid forwarding.
As a temporary workaround, we are now storing which pid is to be used for a given WebProcess in RemoteMediaSessionHelperProxy.
When starting to produce data for camera, we are then ensuring that this pid is correctly set whenever starting to produce data.

We allow overriding of pid in MediaSessionHelperiOS, by adding a new ShouldOverride boolean.
Only AVVideoCaptureSource is using ShouldOverride::Yes, which should limit potential fallouts to camera capture cases.

* Source/WebCore/platform/audio/ios/MediaSessionHelperIOS.h:
* Source/WebCore/platform/audio/ios/MediaSessionHelperIOS.mm:
(MediaSessionHelperiOS::providePresentingApplicationPID):
* Source/WebKit/GPUProcess/GPUConnectionToWebProcess.cpp:
(WebKit::GPUConnectionToWebProcess::overridePresentingApplicationPIDIfNeeded):
* Source/WebKit/GPUProcess/GPUConnectionToWebProcess.h:
* Source/WebKit/GPUProcess/media/ios/RemoteMediaSessionHelperProxy.cpp:
(WebKit::RemoteMediaSessionHelperProxy::providePresentingApplicationPID):
(WebKit::RemoteMediaSessionHelperProxy::overridePresentingApplicationPIDIfNeeded):
* Source/WebKit/GPUProcess/media/ios/RemoteMediaSessionHelperProxy.h:
* Source/WebKit/UIProcess/Cocoa/UserMediaCaptureManagerProxy.cpp:
(WebKit::UserMediaCaptureManagerProxy::startProducingData):
* Source/WebKit/UIProcess/Cocoa/UserMediaCaptureManagerProxy.h:
(WebKit::UserMediaCaptureManagerProxy::ConnectionProxy::startProducingData):
* Source/WebKit/WebProcess/GPU/media/ios/RemoteMediaSessionHelper.cpp:
(WebKit::RemoteMediaSessionHelper::providePresentingApplicationPID):
* Source/WebKit/WebProcess/GPU/media/ios/RemoteMediaSessionHelper.h:

Canonical link: https://commits.webkit.org/265986@main

jernoble commented on Source/WebCore/platform/audio/ios/MediaSessionHelperIOS.mm:
```diff
@@ -249,10 +249,10 @@ - (void)stopMonitoringAirPlayRoutes;
 {
 #if HAVE(CELESTIAL)
-    if (m_presentedApplicationPID)
+    if (m_presentedApplicationPID && (*m_presentedApplicationPID == pid || shouldOverride == ShouldOverride::No))
```
This seems unnecessary. Why do we need the `shouldOverride` flag? Can't we just do something like:

```
if (m_presentedApplicationPID == pid)
    return;
```

jernoble replied:
It seems like there's a bunch of missing logic here. Something in the GPU process needs to decide which of the GPUConnectionToWebProcess should provide the "presented application PID". Right now, each connection just "steals" the presentingApplicationPID setting whenever it starts capture. That _kind of_ works now but only because of policy. It seems like there should be some code inside GPUProcess that is tracking which WebContent process is capturing and doing some kind of MRU tracking.

youennf replied:
> This seems unnecessary. Why do we need the `shouldOverride` flag?

It reduces the potential impact of this patch to capture only.
I agree we should remove shouldOverride at some point but this does not seem to be the right time.

> Right now, each connection just "steals" the presentingApplicationPID setting whenever it starts capture. That _kind of_ works now but only because of policy.

Capture can only happen one page/one process at a time, this is enforced at UIProcess level.
If we change this policy, this patch would break but many things would break at the same time.

This patch is a temporary workaround.
The long term plan is to have the capture session instance able to track which application it is related to with an audit token (or something similar). This patch would then become moot when we will have this.

jernoble replied:
Okay, lets prioritize cleaning this up soon after this workaround lands.
